concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,2111 @@
+        K: Serialize + Ord,
+        V: Serial + DeserialWithState<StateApi> + Deletable, {
+        let v = self.key_value.remove_and_get(key);
+        if v.is_some() && !self.key_order.remove(key) {
+            // Inconsistency between the map and ordered_set.
+            crate::trap();
+        }
+        v
+    }
+
+    /// Remove a key from the map.
+    /// This also deletes the value in the state.
+    pub fn remove(&mut self, key: &K)
+    where
+        K: Serialize + Ord,
+        V: Serial + DeserialWithState<StateApi> + Deletable, {
+        if self.key_order.remove(key) {
+            self.key_value.remove(key);
+        }
+    }
+
+    /// Get a reference to the value corresponding to the key.
+    pub fn get(&self, key: &K) -> Option<StateRef<V>>
+    where
+        K: Serialize,
+        V: Serial + DeserialWithState<StateApi>, {
+        // Minor optimization by first checking whether the collection is empty, since
+        // this information is kept at the root of the ordered set it is already loaded
+        // into memory. In the case of the empty collection we can then save a key
+        // lookup.
+        if self.key_order.is_empty() {
+            None
+        } else {
+            self.key_value.get(key)
+        }
+    }
+
+    /// Get a mutable reference to the value corresponding to the key.
+    pub fn get_mut(&mut self, key: &K) -> Option<StateRefMut<V, StateApi>>
+    where
+        K: Serialize,
+        V: Serial + DeserialWithState<StateApi>, {
+        // Minor optimization by first checking whether the collection is empty, since
+        // this information is kept at the root of the ordered set it is already loaded
+        // into memory. In the case of the empty collection we can then save a key
+        // lookup.
+        if self.key_order.is_empty() {
+            None
+        } else {
+            self.key_value.get_mut(key)
+        }
+    }
+
+    /// Returns `true` if the map contains a value for the specified key.
+    pub fn contains_key(&self, key: &K) -> bool
+    where
+        K: Serialize + Ord, {
+        self.key_order.contains(key)
+    }
+
+    /// Returns the smallest key in the map, which is strictly larger than the
+    /// provided key. `None` meaning no such key is present in the map.
+    pub fn higher(&self, key: &K) -> Option<StateRef<K>>
+    where
+        K: Serialize + Ord, {
+        self.key_order.higher(key)
+    }
+
+    /// Returns the smallest key in the map, which is equal or larger than the
+    /// provided key. `None` meaning no such key is present in the map.
+    pub fn eq_or_higher(&self, key: &K) -> Option<StateRef<K>>
+    where
+        K: Serialize + Ord, {
+        self.key_order.eq_or_higher(key)
+    }
+
+    /// Returns the largest key in the map, which is strictly smaller than the
+    /// provided key. `None` meaning no such key is present in the map.
+    pub fn lower(&self, key: &K) -> Option<StateRef<K>>
+    where
+        K: Serialize + Ord, {
```

Comment:
I did that at first, but I find this more readable in the documentation and code, since the constraints are next to each function rather than grouped somewhere above the function description.
Also, not every function depends on these traits.